src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,551 @@
+	 */
+	cd->direction.max_lag = Q_MULTSR_32X32(fs, t_max, 0, 15, 0) + 1;
+	n = (cd->max_frames + (2 * cd->direction.max_lag + 1)) * ch_count;
+	cd->direction.d_size =  n * sizeof(int16_t);
+	cd->direction.d = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, cd->direction.d_size);
+	if (!cd->direction.d) {
+		rfree(cd->direction.df2t_delay);
+		return -EINVAL;
+	}
+
+	/* Set needed pointers to xcorr delay line, advance write pointer by max_lag to keep read
+	 * always behind write
+	 */
+	cd->direction.d_end = cd->direction.d + n;
+	cd->direction.wp = cd->direction.d;
+	cd->direction.rp = cd->direction.wp;
+	cd->direction.wp += ch_count * (cd->direction.max_lag + 1);
+
+	/* xcorr result is temporary but too large for stack so it is allocated here */
+	cd->direction.r_size = (2 * cd->direction.max_lag + 1) * sizeof(int32_t);
+	cd->direction.r = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, cd->direction.r_size);
+	if (!cd->direction.r) {
+		rfree(cd->direction.df2t_delay);
+		rfree(cd->direction.d);
+		return -EINVAL;
+	}
+
+	/* Check for line array mode */
+	cd->direction.line_array = line_array_mode_check(cd);
+
+	/* Initialize direction to zero radians, set initial step sign to +1 */
+	cd->direction.az = 0;
+	cd->direction.step_sign = 1;
+	return 0;
+}
+
+void tdfb_direction_free(struct tdfb_comp_data *cd)
+{
+	rfree(cd->direction.df2t_delay);
+	rfree(cd->direction.d);
+	rfree(cd->direction.r);
+}
+
+/* Measure level of one channel */
+static inline void level_update(struct tdfb_comp_data *cd, int frames, int ch_count, int channel)
+{
+	int64_t tmp = 0;
+	int64_t ambient;
+	int64_t level;
+	int16_t s;
+	int16_t *p = cd->direction.rp + channel;
+	int thr;
+	int n;
+
+	/* Calculate mean square level */
+	for (n = 0; n < frames; n++) {
+		s = *p;
+		p += ch_count;
+		tdfb_cinc_s16(&p, cd->direction.d_end, cd->direction.d_size);
+		tmp += ((int64_t)s * s);
+	}
+
+	/* Track level times 256 for better fast/slow filter precision */
+	cd->direction.level = sat_int32(tmp / frames);
+
+	/* Slow track minimum level and use it as ambient noise level estimate */
+	level = ((int64_t)cd->direction.level) << 32;
+	ambient = cd->direction.level_ambient;
+	if (level < ambient)
+		ambient = (level >> FAST_LEVEL_SHIFT) - (ambient >> FAST_LEVEL_SHIFT) + ambient;
+	else
+		ambient = (level >> SLOW_LEVEL_SHIFT) - (ambient >> SLOW_LEVEL_SHIFT) + ambient;
+
+	thr = sat_int32(Q_MULTS_32X32(ambient >> 32, POWER_THRESHOLD, 31, 10, 31));
+	cd->direction.level_ambient = ambient;
+	cd->direction.trigger = cd->direction.trigger << 1;
+	if (cd->direction.level > thr)
+		cd->direction.trigger |= 1;
+
+	/* Levels update runs always when tracking, increment frame counter until wrap */
+	cd->direction.frame_count_since_control += frames;
+	if (cd->direction.frame_count_since_control < 0)
+		cd->direction.frame_count_since_control = INT32_MAX;
+}
+
+static inline int find_max_value_index(int32_t *v, int length)
+{
+	int32_t max_val = v[0];
```

Comment:
nit: if you are storing the index then you don't need this variable

Reply:
Yep, fixed!